### PR TITLE
feat: Graph based deploy

### PR DESCRIPTION
**Pull request by Laubi**
#### What this PR does / Why we need it:
This PR introduces an innovative approach to deploying Monaco configurations. To enable the new functionality, the Feature Flag `MONACO_FEAT_GRAPH_DEPLOY` must be enabled.

The novel deployment method divides the configurations to be deployed into independent components. Each component is then deployed separately without sharing information between them.

By doing so, we can avoid deploying the children of configurations that have failed to be deployed or were skipped. However, configurations within a component that don't depend on failed configurations will still be deployed.

#### Does this PR introduce a user-facing change?
Upon activating this new feature, instead of displaying errors regarding missing references, the system will indicate that certain configurations have been skipped for deployment due to parent failures. This enhancement enhances the clarity of log messages and improves the understanding of deployment failures.

### CODE REVIEW

**Comment by UnseenWizzard** [comment]
@Laubi I retriggered the E2E tests - failure seems unrelated to the changes :/ 

> 2023-07-19T10:02:21Z	warn	API for "management-zone" is deprecated! Please consider migrating to "builtin:management-zones"!
> **2023-07-19T10:02:22Z	debug		Created new object for mzone-1_20230719100220_6913_allconfigs (2142474705747721855)**
> 2023-07-19T10:02:22Z	info	Deployed project:management-zone:zone_20230719100220_6913_allconfigs successfully (graph-node-id 62)
> 2023-07-19T10:02:22Z	info	Deploying config project:maintenance-window:maintenancewindow_20230719100220_6913_allconfigs
> 2023-07-19T10:02:22Z	warn	API for "maintenance-window" is deprecated! Please consider migrating to "builtin:alerting.maintenance-window"!
> 2023-07-19T10:02:22Z	error	Failed to deploy config project:maintenance-window:maintenancewindow_20230719100220_6913_allconfigs: Failed to create DT object My First Maintenance Window_20230719100220_6913_allconfigs (HTTP 400)!
>     Response was: {"error":{"code":400,"message":"Constraints violated.","constraintViolations":[{"path":"builtin:alerting.maintenance-window/0/filters/0/managementZones/0","message":"Given property 'managementZones/0' with value: '2142474705747721855' is not a valid value in datasource.","parameterLocation":"PAYLOAD_BODY","location":null}]}}
> 2023-07-19T10:02:22Z	error	=== Deployment Error Details ===
> 2023-07-19T10:02:22Z	error	platform_env(default) project:maintenance-window:maintenancewindow_20230719100220_6913_allconfigs deploy.configDeployErr Failed to create DT object My First Maintenance Window_20230719100220_6913_allconfigs (HTTP 400)!
>     Response was: {"error":{"code":400,"message":"Constraints violated.","constraintViolations":[{"path":"builtin:alerting.maintenance-window/0/filters/0/managementZones/0","message":"**Given property 'managementZones/0' with value: '2142474705747721855' is not a valid value in datasource."**,"parameterLocation":"PAYLOAD_BODY","location":null}]}}
> Error: errors during Deployment

File: pkg/deploy/deploy_test/deploy_graph_test.go [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+	c := deploy.EnvironmentClients{
+		deploy.EnvironmentInfo{Name: "env"}: deploy.DummyClientSet,
+	}
+
+	t.Run("deployment error - stop on error", func(t *testing.T) {
+		errors := deploy.DeployConfigGraph(p, c, deploy.DeployConfigsOptions{})
+		assert.Equal(t, 1, len(errors), fmt.Sprintf("Expected 1 error, but just got %d", len(errors)))
+	})
+
+	t.Run("deployment error - stop on error", func(t *testing.T) {
```
Comment (Laubi) [inline pkg/deploy/deploy_test/deploy_graph_test.go:412]: Duplicated test name

File: cmd/monaco/deploy/deploy.go [hunk 2]
```diff
@@ -58,53 +60,65 @@ func deployConfigs(fs afero.Fs, manifestPath string, environmentGroups []string,
...
-			break
+	if featureflags.DependencyGraphBasedDeploy().Enabled() {
+		clientSets, err := createDeployClientSets(loadedManifest.Environments, dryRun)
+		if err != nil {
+			return fmt.Errorf("failed to create API clients: %w", err)
+		}
+		environmentDeployErrs := deploy.DeployConfigGraph(filteredProjects, clientSets, deploy.DeployConfigsOptions{
+			ContinueOnErr:  continueOnErr,
+			DryRun:         dryRun,
+			SupportArchive: support.SupportArchive,
```
Comment (Laubi) [inline cmd/monaco/deploy/deploy.go:78]: Throw out

File: pkg/graph/graph.go [hunk 3]
```diff
@@ -80,24 +82,29 @@ func (graphs ConfigGraphPerEnvironment) SortConfigs(environment string) ([]confi
...
 	sortedCfgs := make([]config.Config, len(sortedNodes))
 	for i, n := range sortedNodes {
-		sortedCfgs[i] = *n.(configNode).config
+		sortedCfgs[i] = *n.(ConfigNode).Config
 	}
 
 	return sortedCfgs, nil
 }
 
+type SortedComponent struct {
```
Comment (Laubi) [inline pkg/graph/graph.go:91]: Write documentation on why the graph is not enough (it would be) but we also have the topology-sorted nodes in

**Comment by UnseenWizzard** [comment]
E2E tests failed with 500/503 errors on cleanup 😢 retriggered

File: pkg/graph/graph_test/graph_test.go
Comment (warber) [inline pkg/graph/graph_test/graph_test.go:1]: please move this file up into the `graph` directory and rename the package name to just `graph_test`

**Review by warber — DISMISSED** [review]
please move the `graph_test.go` file up into the `graph` directory and rename the package name to just `graph_test`

**Review by warber — APPROVED** [review]

**Pull request merged**